src/GridFS/GridFsUpload.php

```diff
@@ -0,0 +1,169 @@
+<?php
+namespace MongoDB\GridFS;
+
+use MongoDB\Collection;
+use MongoDB\Exception\RuntimeException;
+use MongoDB\Exception\UnexpectedTypeException;
+use MongoDB\Exception\InvalidArgumentException;
+use MongoDB\BSON;
+/**
+ * GridFsupload abstracts the processes of inserting into a GridFSBucket
+ *
+ * @api
+ */
+class GridFsUpload extends GridFsStream
+{
+    private $ctx;
+    private $bufferLength;
+    private $indexChecker;
+    private $length=0;
+    /**
+     * Constructs a GridFS upload stream
+     *
+     * Supported options:
+     *
+     *  * contentType (string): DEPRECATED content type to be stored with the file.
+     *    This information should now be added to the metadata
+     *
+     *  * aliases (array of strings): DEPRECATED An array of aliases.
+     *    Applications wishing to store aliases should add an aliases field to the
+     *    metadata document instead.
+     *
+     *  * metadata (array or object): User data for the 'metadata' field of the files
+     *    collection document.
+     *
+     *  * writeConcern (MongoDB\Driver\WriteConcern): Write concern.
+     *
+     * @param \MongoDB\Collection    $filesCollection  Files Collection
+     * @param \MongoDB\Collection    $chunksCollection Chunks Collection
+     * @param int32                  $chunkSizeBytes   Size of chunk
+     * @param string                 $filename         Filename to insert
+     * @param array                  $options          File options
+     * @throws InvalidArgumentException
+     */
+    public function __construct(
+        Bucket $bucket,
```

Comment:
So after extensive toying around I have found that this still suffers from the main flaw of the fopen call behaving as a wall. In order for the ensure indexes [call to happen at the time the spec insists that it should] (https://github.com/mongodb/specifications/blob/master/source/gridfs/gridfs-spec.rst#before-write-operations) an object with those functions **must** be passed through to the stream. It cannot occur in the bucket class because the bucket returns a stream in the `openUploadStream` function and has no understanding of what is happening behind that wall. It seems quite reasonable to me to continue passing a wrapper with the collections around. It doesn't seem to me that it would introduce too much overhead as it is just copying a pointer, so copying a pointer to a wrapper is just as much work as copying a pointer to a collection.